src/payload/addr.rs

```diff
@@ -0,0 +1,711 @@
+//! IP address resources.
+
+use std::{error, fmt};
+use std::net::{AddrParseError, IpAddr, Ipv4Addr, Ipv6Addr};
+use std::num::ParseIntError;
+use std::str::FromStr;
+#[cfg(feature = "serde")] use serde::{
+    Deserialize, Deserializer, Serialize, Serializer
+};
+
+
+//------------ Bits ----------------------------------------------------------
+
+/// The value of an IP address.
+///
+/// This private type holds the content of an IP address. It can be both an
+/// IPv4 and IPv6 address. It keeps the address internally as a 128 bit
+/// unsigned integer. IPv6 address are kept in all bits host byte order while
+/// IPv4 addresses are kept in the upper four bytes and are padded with zero
+/// bits. This makes it possible to count prefix lengths the same way for both
+/// addresses, i.e., starting from the top of the raw integer.
+///
+/// There is no way of distinguishing between IPv4 and IPv6 from just a value
+/// of this type. This information needs to be carried separatedly.
+#[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
+struct Bits(u128);
+
+impl Bits {
+    /// Creates a new address from 128 raw bits in host byte order.
+    pub fn new(bits: u128) -> Self {
+        Bits(bits)
+    }
+
+    /// Creates a new address value for an IPv4 address.
+    pub fn from_v4(addr: Ipv4Addr) -> Self {
+        Self::new(u128::from(u32::from(addr)) << 96)
+    }
+
+    /// Creates a new address value for an IPv4 address.
+    pub fn from_v6(addr: Ipv6Addr) -> Self {
+        Self::new(u128::from(addr))
+    }
+
+    /// Returns the raw bits of the underlying integer.
+    pub fn into_int(self) -> u128 {
```

Comment:
The term ‘int’ here was intended to mean ‘the integer type representing the bits’ rather than ‘what C calls an `int`.’ Given that this is an private type, I think this is fine.